### PR TITLE
Added a recursive look-up mechanism for the java.lang.ClassLoader.defineClass() native method and the necessary unit test testDefineClass for this change (#199)

### PR DESCRIPTION
**Pull request by rtrimana**
1) The recursive look-up is attempted just in the native method `defineClass()` in the model class `JPF_java_lang_ClassLoader` through exception catching and handling. The look-up is attempted in the `SystemClassLoader` if the current `ClassLoader` fails to find the required classes. If the class still could not be found, the method would then throw a JPF's `java.lang.ClassNotFoundException` properly.

2) When the `ClassInfo` has been defined, we just return the necessary object reference number and not return a `LinkageError` exception as what it did before.

3) I have also created a unit test for the `defineClass()` method. All the unit tests have been run and they successfully passed.

4) There are a few minor changes in the build.gradle and jpf.properties files to download the necessary ASM library (for class writing), and get the right `classpath` for it.

### CODE REVIEW

**Comment by cyrille-artho** [comment]
Thanks, the patch is almost ready to be merged. I just have a few minor comments, the most important one being about avoiding a dependency for the main build if this is possible. Please see my comments in the review.

**Comment by rtrimana** [comment]
Thank you but I don't see your review/comments in here https://github.com/javapathfinder/jpf-core/pull/203/files. And the "Reviewers" list on the right-hand side of this page also shows "No reviews".

Did you probably forget to submit the review or perhaps I missed something here? :)

**Comment by cyrille-artho** [comment]
Sorry, I forgot to submit the review. Here it is.

File: build.gradle [hunk 1]
```diff
@@ -20,6 +20,12 @@ buildscript {
 
 dependencies {
     testImplementation "junit:junit:4.12"
+    compile "org.ow2.asm:asm:7.1"
```
[the remark this answers is not in the document]
Comment (rtrimana) [reply build.gradle]: Alright, I am going to change it to "testCompile".

File: build.gradle [hunk 2]
```diff
@@ -203,6 +209,7 @@ task buildJars {
     dependsOn createJpfJar
     dependsOn createRunJpfJar
     dependsOn createRunTestJar
+	dependsOn copyLibs
```
[the remark this answers is not in the document]
Comment (rtrimana) [reply build.gradle]: Thanks for catching this. Apparently my editor was pre-configured with a different indentation style.

File: src/peers/gov/nasa/jpf/vm/JPF_java_lang_ClassLoader.java [hunk 3]
```diff
@@ -134,33 +134,50 @@ public int findSystemClass__Ljava_lang_String_2__Ljava_lang_Class_2 (MJIEnv env,
     String cname = env.getStringObject(nameRef);
     ClassLoaderInfo cl = env.getClassLoaderInfo(objRef);
 
-    // determine whether that the corresponding class is already defined by this 
-    // classloader, if so, this attempt is invalid, and loading throws a LinkageError
+    // determine whether that the corresponding class is already defined by this
+    // classloader, if so, just return the already defined class
```
[the remark this answers is not in the document]
Comment (rtrimana) [reply src/peers/gov/nasa/jpf/vm/JPF_java_lang_ClassLoader.java]: This is a good catch and definitely was my mistake. I was assuming that the defineClass() method should have returned the necessary ClassInfo object when there is already one. Indeed the standard JVM will return the same java.lang.LinkageError when this method is called twice to define the same class in the same ClassLoader.

Therefore, I am changing the patch. Basically we keep checking in the SystemClassLoaderInfo when we cannot find a class and if possible we define the class through the SystemClassLoaderInfo.

**Comment by cyrille-artho** [comment]
Looks good, thanks!

**Comment by cyrille-artho** [comment]
Merged into master.
I used a squash-commit to merge all intermediate changes into a single commit, and removed parts of the commit message that referred to commits that undid earlier changes, as these no longer appear in the final, squashed commit.

**Pull request merged**